### PR TITLE
Run context condition checks only once, after model parsing

### PR DESCRIPTION
**Pull request by clinssen**
This is a spin-off from #1050.

Refactor cocos such that they run only once, after model parsing.

Fixes some other bugs that emerged as part of this refactor. For instance, two cocos turned out to actually make needed changes to the model; the modification part has now been spun off to the new AssignImplicitConversionFactorsTransformer. (I can also make the transformer into a pure visitor; please let me know if you prefer this approach.)

Changes order of operations in compartmental code generator: to prevent symbol not found error, add propagator variable first, then add the update expressions that reference this variable.

### CODE REVIEW

File: pynestml/codegeneration/nest_code_generator.py [hunk 1]
```diff
@@ -908,8 +910,8 @@ def update_symbol_table(self, neuron) -> None:
         """
         SymbolTable.delete_model_scope(neuron.get_name())
         symbol_table_visitor = ASTSymbolTableVisitor()
-        symbol_table_visitor.after_ast_rewrite_ = True
         neuron.accept(symbol_table_visitor)
+        CoCosManager.check_cocos(neuron, after_ast_rewrite=True)
```
Comment (pnbabu) [inline pynestml/codegeneration/nest_code_generator.py:914]: Should the cococs be checked inside each code generator? Shouldn't it rather happen between transformation and code generation stages?

File: pynestml/transformers/assign_implicit_conversion_factors_transformer.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+from pynestml.utils.type_caster import TypeCaster
+from pynestml.visitors.ast_visitor import ASTVisitor
+
+
+class AssignImplicitConversionFactorsTransformer(Transformer):
+    r"""
+    Assign implicit conversion factors in expressions.
+    """
+
+    def transform(self, models: Union[ASTNode, Sequence[ASTNode]]) -> Union[ASTNode, Sequence[ASTNode]]:
```
Comment (pnbabu) [inline pynestml/transformers/assign_implicit_conversion_factors_transformer.py:50]: Which are the two cocos that are replaced by this transformer? Do they also have to be removed from the cocos manager class?

**Review by pnbabu — CHANGES_REQUESTED** [review]

File: pynestml/codegeneration/nest_code_generator.py (hunk 1 above)
Comment (clinssen) [reply pynestml/codegeneration/nest_code_generator.py:914]: Indeed, the primary coco checking call (to ``CoCosManager.check_cococs()``) is in the frontend. This is an additional call (with ``after_ast_rewrite=True``) after model transformations, as an extra sanity check.

File: pynestml/transformers/assign_implicit_conversion_factors_transformer.py (hunk 2 above)
Comment (clinssen) [reply pynestml/transformers/assign_implicit_conversion_factors_transformer.py:50]: Thank you for remarking upon this. The calls  have been effectively removed by setting ``set_implicit_conversion_factor_on_lhs`` to False on https://github.com/nest/nestml/pull/1105/files#diff-e13518a54108e0f87db406aee6c43eadbebc0bb6908943a1f045136bc1c14024R43. This affects ``co_co_illegal_expression``, ``co_co_user_defined_function_correctly_defined`` and ``co_co_function_calls_consistent``.

File: pynestml/codegeneration/spinnaker_code_generator.py [hunk 3]
```diff
@@ -216,6 +215,7 @@ def generate_code(self, models: Sequence[ASTModel]) -> None:
         for model in models:
```
Comment (pnbabu) [inline pynestml/codegeneration/spinnaker_code_generator.py:215]: Can these two for loops be merged into one?

File: pynestml/transformers/assign_implicit_conversion_factors_transformer.py (hunk 2 above)
Comment (pnbabu) [reply pynestml/transformers/assign_implicit_conversion_factors_transformer.py:50]: Can these cocos be removed then?

File: pynestml/codegeneration/spinnaker_code_generator.py (hunk 3 above)
Comment (clinssen) [reply pynestml/codegeneration/spinnaker_code_generator.py:215]: Done!

File: pynestml/utils/model_parser.py [hunk 4]
```diff
@@ -142,10 +144,14 @@ def parse_file(cls, file_path=None):
         for model in ast.get_model_list():
             model.accept(ASTSymbolTableVisitor())
             SymbolTable.add_model_scope(model.get_name(), model.get_scope())
+            Logger.set_current_node(model)
+            AssignImplicitConversionFactorsTransformer().transform(model)
```
[the remark this answers is not in the document]
Comment (clinssen) [reply pynestml/utils/model_parser.py]: Done! Changed the transformer into a visitor.

**Review by pnbabu — APPROVED** [review]
Looks good to me, thank you!

**Comment by clinssen** [comment]
Thank you for the review! Let's merge as soon as the tests pass.

**Pull request merged**